### ISSUE TITLE
WDIO Test Runner: browser.saveScreenshot() hangs/times out

### ISSUE THREAD

**Issue by pagemedias** [webdriverio/webdriverio#1557]
## The problem

I am running WDIO Test Runner from TeamCity and every time I try to save a screenshot WDIO hangs (e.g. build time 6mins instead of 40secs) and results in timeout.

## Environment

* wdio version: 0.3.3
* Node.js version: 6.3.1
* wdio testrunner
* Additional wdio packages used (if applicable):
 * wdio-jasmine-framework
 * wdio-junit-reporter
 * wdio-teamcity-reporter
 * wdio-selenium-standalone-service

## Details

I am running WDIO Test Runner from TeamCity and every time I try to save a screenshot...
```
browser.saveScreenshot('./myscreenshot.png');
```
WDIO hangs (e.g. build time 6mins instead of 40secs) and results in:
```
Error: Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL.
```

If I remote into TeamCity server (with AD account) and run WDIO from command line (same command as TeamCity runs) then it creates the screenshot fine.

## Link to Selenium/WebdriverIO logs
https://gist.github.com/pagemedias/a1ba0e49147867e28e7412b19484e9e7

**Comment by christian-bromann** [comment]
try to wrap the save screenshot in a try catch (I guess you use sync mode) and see if the catch part gets invoked with an error.

**Comment by pagemedias** [comment]
Thanks for the reply @christian-bromann I just tried removing `browser.screenshot()` and instead adding `screenshotPath: './errorShots',` to wdio.conf.js and I got this error:

```
Failed to take screenshot on reject: 	 {"message":"Couldn't connect to selenium server","type":"RuntimeError","seleniumStack":{"status":-1,"type":"ECONNREFUSED","message":"Couldn't connect to selenium server","orgStatusMessage":"Couldn't connect to selenium server"}}
```

Weird thing is that my spec passes fine (so obviously able to connect to selenium). Just taking screenshots that is the issue

**Comment by pagemedias** [comment]
tried wrapping `browser.saveScreenshot()` in try catch (yes sync mode) and returned same error:

```
RuntimeError: Couldn't connect to selenium server
[16:04:15]  [Step 3/3]     at Object.Future.wait (C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\fibers\future.js:449:15)
[16:04:15]  [Step 3/3]     at Object.saveScreenshot (C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\wdio-sync\build\index.js:414:27)
[16:04:15]  [Step 3/3]     at Object.<anonymous> (C:\TeamCity\buildAgent\work\3169373566d171fe\spec\ui\valves-spec.js:25:21)
[16:04:15]  [Step 3/3]     at C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\wdio-sync\build\index.js:642:26
[16:04:15]  [Step 3/3]     at Object.executeSync (C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\wdio-sync\build\index.js:640:12)
[16:04:15]  [Step 3/3]     at C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\wdio-sync\build\index.js:877:34
[16:04:15]  [Step 3/3]     at session("delete") - C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\webdriverio\build\lib\commands\end.js:26:17
[16:04:15]  [Step 3/3]     at end() - C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\webdriverio\build\lib\runner.js:281:45
[16:04:15]  [Step 3/3]   message: 'Couldn\'t connect to selenium server',
[16:04:15]  [Step 3/3]   type: 'RuntimeError',
[16:04:15]  [Step 3/3]   seleniumStack: 
[16:04:15]  [Step 3/3]    { status: -1,
[16:04:15]  [Step 3/3]      type: 'ECONNREFUSED',
[16:04:15]  [Step 3/3]      message: 'Couldn\'t connect to selenium server',
[16:04:15]  [Step 3/3]      orgStatusMessage: 'Couldn\'t connect to selenium server' }
```

**Comment by pagemedias** [comment]
Any tips on where to star to debug this? Thanks!

**Comment by christian-bromann** [comment]
Well as the error says you call saveScreenshot after you closed the selenium session. Do you call it in the onCompete hook?

**Comment by pagemedias** [comment]
Call it in jasmineNodeOpts: {expectationResultHandler:{}} just like example on http://webdriver.io/guide/testrunner/frameworks.html

Ok thanks. Do you mean I am doing browser.end() on afterAll() in spec?

If I don't close the selenium session in afterAll(), will wdio runner do it automatically?

Thanks heaps!

**Comment by christian-bromann** [comment]
> will wdio runner do it automatically?

yes, you don't need to call it

Try to use afterStep instead of the expectation result handler .. seems that it gets called after the session ends

**Comment by pagemedias** [comment]
I removed the `browser.end()` altogether from spec and tried adding `browser.saveScreenshot()` to `after: function (result, capabilities, specs) {}` and directly to spec and get the same error. I also tested running `browser.title()` to check that selenium still running and works (gets correct title).

```
[23:13:34]  COMMAND  GET      "/wd/hub/session/1e9d29b0-3679-4626-af44-c4dce4a9b09f/title"
[23:13:34]  DATA     {}
[23:13:34]  RESULT       "My Website"
{ state: 'success',
  sessionId: '1e9d29b0-3679-4626-af44-c4dce4a9b09f',
  hCode: 311791386,
  value: 'My Website',
  class: 'org.openqa.selenium.remote.Response',
  status: 0 }
[23:13:34]  COMMAND  GET      "/wd/hub/session/1e9d29b0-3679-4626-af44-c4dce4a9b09f/screenshot"
[23:13:34]  DATA     {}
C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\fibers\future.js:313
                     throw(ex);
                     ^

RuntimeError: Couldn't connect to selenium server
    at Object.Future.wait (C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\fibers\future.js:449:15)
    at Object.saveScreenshot (C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\wdio-sync\build\index.js:371:31)
    at Object.exports.config.after (C:\TeamCity\buildAgent\work\3169373566d171fe\wdio.conf.js:206:17)
    at C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\wdio-sync\build\index.js:269:34
    - - - - -
    at screenshot() - C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\webdriverio\build\lib\commands\saveScreenshot.js:16:17
    at saveScreenshot("./testing.png") - C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\wdio-sync\build\index.js:364:37
    at title() - C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\wdio-sync\build\index.js:364:37
    at url("https://example.com/mysite") -     at C:\TeamCity\buildAgent\work\3169373566d171fe\node_modules\wdio-sync\build\index.js:392:3
```

**Comment by pagemedias** [comment]
I did some more testing and got it working (taking screenshots) with using npm selenium-standalone instead of wdio-selenium-standalone-service.

**i.e.**
I commented out `services: ['selenium-standalone'],` in wdio.conf

In separate command line window I run:
```
npm selenium-standalone install
npm selenium-standalone start
```

Then run TeamCity build.

I would prefer to use wdio-selenium-standalone-service, because now I need to factor in npm selenium-standalone dependency (e.g. run npm selenium-standalone install and npm selenium-standalone start every time before I run \node_modules\.bin\wdio)
